modules/contact/test/tests/mortar_restart/tests

```diff
@@ -25,5 +25,7 @@
     prereq = 'frictional_bouncing_block_action_restart_1'
     valgrind = 'heavy'
     method = '!dbg'
+    # skip test if test is being run out-of-tree. Issue Ref: #25279
+    installation_type = in_tree
```

Comment:
It would be nice to understand what are criteria for tests being run out of tree. I don't know why this one won't run out of tree. Is it because there's a prereq? Because it requires a file mesh?

Reply:
I'd be happy to add real reasons rather than blanket ones... but I don't know much about what generates these errors. Here is the produced error for `2d-finite-strain-steady/tests`:

```pre
2d-finite-strain-steady.test: *** ERROR ***
2d-finite-strain-steady.test: ADFParser::JITCompile() failed. Evaluation not possible.
2d-finite-strain-steady.test: 
2d-finite-strain-steady.test: Stack frames: 19
2d-finite-strain-steady.test: 0: 0   libmesh_opt.0.dylib                 0x0000000108827ce0 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char>>&) + 1044
2d-finite-strain-steady.test: 1: 1   libmoose-opt.0.dylib                0x00000001050aff58 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) + 680
2d-finite-strain-steady.test: 2: 2   libmoose-opt.0.dylib                0x000000010437df44 void mooseError<char const (&) [57]>(char const (&) [57]) + 216
2d-finite-strain-steady.test: 3: 3   libmoose-opt.0.dylib                0x0000000104ec2394 ADFParser::JITCompile() + 948
2d-finite-strain-steady.test: 4: 4   libmoose-opt.0.dylib                0x000000010441899c ParsedMaterialHelper<true>::functionsOptimize() + 48
2d-finite-strain-steady.test: 5: 5   libmoose-opt.0.dylib                0x000000010441af48 ParsedMaterialHelper<true>::functionsPostParse() + 28
2d-finite-strain-steady.test: 6: 6   libmoose-opt.0.dylib                0x000000010441adc8 ParsedMaterialHelper<true>::functionParse(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<PostprocessorName, std::__1::allocator<PostprocessorName>> const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, std::__1::vector<double, std::__1::allocator<double>> const&) + 1988
2d-finite-strain-steady.test: 7: 7   libmoose-opt.0.dylib                0x0000000104463844 ParsedMaterialTempl<true>::ParsedMaterialTempl(InputParameters const&) + 480
2d-finite-strain-steady.test: 8: 8   libmoose-opt.0.dylib                0x00000001044635e8 RegistryEntry<ParsedMaterialTempl<true>>::build(InputParameters const&) + 64
2d-finite-strain-steady.test: 9: 9   libmoose-opt.0.dylib                0x0000000105031b84 Factory::create(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters const&, unsigned int, bool) + 344
2d-finite-strain-steady.test: 10: 10  libmoose-opt.0.dylib                0x0000000104815990 std::__1::shared_ptr<MaterialBase> Factory::create<MaterialBase>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters const&, unsigned int) + 56
2d-finite-strain-steady.test: 11: 11  libmoose-opt.0.dylib                0x0000000104814268 FEProblemBase::addMaterialHelper(std::__1::vector<MaterialWarehouse*, std::__1::allocator<MaterialWarehouse*>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters&) + 564
2d-finite-strain-steady.test: 12: 12  libmoose-opt.0.dylib                0x0000000104813fd8 FEProblemBase::addMaterial(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, InputParameters&) + 112
2d-finite-strain-steady.test: 13: 13  libmoose-opt.0.dylib                0x0000000104b37de0 ActionWarehouse::executeActionsWithAction(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&) + 1160
2d-finite-strain-steady.test: 14: 14  libmoose-opt.0.dylib                0x0000000104b659bc ActionWarehouse::executeAllActions() + 244
2d-finite-strain-steady.test: 15: 15  libmoose-opt.0.dylib                0x0000000105087e58 MooseApp::runInputFile() + 152
2d-finite-strain-steady.test: 16: 16  libmoose-opt.0.dylib                0x000000010508468c MooseApp::run() + 1364
2d-finite-strain-steady.test: 17: 17  combined-opt                        0x0000000100737b4c main + 164
2d-finite-strain-steady.test: 18: 18  dyld                                0x000000019f16ff28 start + 2236
2d-finite-strain-steady.test: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
2d-finite-strain-steady.test: [unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
2d-finite-strain-steady.test: :
2d-finite-strain-steady.test: system msg for write_line failure : Bad file descriptor
2d-finite-strain-steady.test: 
2d-finite-strain-steady.test: 
2d-finite-strain-steady.test: Exit Code: 1
2d-finite-strain-steady.test: ################################################################################
2d-finite-strain-steady.test: Tester failed, reason: ERRMSG
```

Reply:
Pinging @dschwen for his perspective on JIT in these cases, or what we might do to allow these tests on all installation types.

Reply:
https://github.com/idaholab/moose/blob/a52b2572d064d7bf7c58825bf243f3f59104eea1/framework/src/utils/ADFParser.C#L50

FParser isn't available in the installed binary it seems?

Reply:
well this is odd:

```pre
    SKIP umat.gap_heat_transfer_umat [test requires "IN_TREE" binary]
<trimmed>
      OK feature_volume_fraction.volume_fraction_test
      OK adaptive_timestepping.test_function_change
      OK combined_plasticity_temperature.temp_dep_yield
      OK restart-transient-from-ss-with-stateful.ss
invOpt_nonlinear.BFGS: Working Directory: /private/var/folders/1j/j662yjdx2zb1gspprj00hd70664rrb/T/tmp.eCRXSczM/combined/combined/invOpt_nonlinear
invOpt_nonlinear.BFGS: Running command: /Users/milljm/moose_installed/share/moose/python/TestHarness/../../../../bin/combined-opt -i main.i Executioner/tao_solver=taolmvm --error --error-override --no-gdb-backtrace
invOpt_nonlinear.BFGS: 
```
I am not sure why `invOpt_nonlinear` is not being skipped like it is supposed to.

Reply:
I don't suppose that JIT compiles (and fails) before evaluating whether or not the test should run or not?

Reply:
damn, the local build on my machine still fails.

Reply:
Well I am at a loss.

```pre
moose                     2023.08.19              build_0    https://conda.software.inl.gov/public
```
...version works. I'll need to dig a little bit to understand what might be different. I am sure this is going to be finicky conda-build environment.

Reply:
I've added https://github.com/idaholab/moose/issues/25293#issuecomment-1692359189. Perhaps we can add reasons as we determine them there

Reply:
I'm a bit late here, but the whole `ADRealMonolithic.h` stuff was created to make JIT compiling work for binary builds. This institutional short term memory loss is a bit concerning. How can we avoid this?

Reply:
FYI, this is documented under modules/phase_field/doc/content/modules/phase_field/FunctionMaterials/JITCompilation.md

Reply:
Resolving to merge a fix for now, but we should continue this discussion

Reply:
> This is a little severe I think.

Ah, I see where the check-in came from :-D. This is not meant as a personal criticism of anyone involved. I'm just observing a structural problem that needs to be addressed IMO. The documentation was quite apparently either not enough, or not discoverable enough. Design docs are supposed to prevent these types of things from happening. 

I suppose we could link explicitly to the relevant design doc from an error message (`ADFParser::JITCompile() failed. Evaluation not possible. Check JITCompilation.md to see if the environment is set up correctly.`).

Reply:
This could be an environment 'thing', but seeing as we've been failing on Civet, where we blow away the environment every time... I gotta ask:

Using `moose` that is available in public _now_, everything works:
```bash
❯ mamba create -n prebuilt moose
❯ mamba activate prebuilt
❯ moose --copy-inputs optimization
❯ cd moose/optimization
❯ moose --run -j 12 --re=nonlinear
Working Directory: /private/tmp/testing/moose/optimization
Running Command: /Users/milljm/mambaforge3/envs/testing/moose/bin/moose_test_runner -j 12 --re nonlinear
      OK executioners/steady_and_adjoint.steady_and_adjoint/nonlinear
      OK executioners/transient_and_adjoint.transient_and_adjoint/nonlinear
      OK optimizationreporter/constant_heat_source.constant_heat_source/nonlinear
      OK optimizationreporter/nonlinear_material.nonlinear_material
--------------------------------------------------------------------------------------------------------------
Ran 4 tests in 1.7 seconds. Average test time 1.0 seconds, maximum test time 1.6 seconds.
```

using `moose` that I am building on HEAD, is failing for the same tests. With or without setting `MOOSE_ADFPARSER_JIT_INCLUDE`.

I'd also like to point out that in public MOOSE as demonstrated above, we are NOT setting `MOOSE_ADFPARSER_JIT_INCLUDE`!

These errors are also only happening on Apple Silicon machines.

Full log of a single test failure will be attached...
[log.txt](https://github.com/idaholab/moose/files/12457657/log.txt)

Reply:
Not sure if it matters, but here is a log of the same test, after setting MOOSE_ADFPARSER_JIT_INCLUDE (still fails)

```pre
Final Test Results:
--------------------------------------------------------------------------------------------------------------
      OK optimizationreporter/constant_heat_source.constant_heat_source/nonlinear
    FAIL optimizationreporter/nonlinear_material.nonlinear_material FAILED (ERRMSG)
    FAIL executioners/transient_and_adjoint.transient_and_adjoint/nonlinear FAILED (ERRMSG)
    FAIL executioners/steady_and_adjoint.steady_and_adjoint/nonlinear FAILED (ERRMSG)
--------------------------------------------------------------------------------------------------------------
Ran 4 tests in 1.4 seconds. Average test time 1.2 seconds, maximum test time 1.3 seconds.
1 passed, 0 skipped, 0 pending, 3 FAILED


*** ERROR ***
Run failed

Stack frames: 7
0: 0   libmesh_opt.0.dylib                 0x0000000108607ce0 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char>>&) + 1044
1: 1   libmoose-opt.0.dylib                0x0000000104e8f940 moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>) + 680
2: 2   libmoose-opt.0.dylib                0x0000000104e5e7f8 void mooseError<char const (&) [11]>(char const (&) [11]) + 216
3: 3   libmoose-opt.0.dylib                0x0000000104e5e410 MooseApp::runInputs() const + 1044
4: 4   libmoose-opt.0.dylib                0x0000000104e63ee0 MooseApp::run() + 960
5: 5   combined-opt                        0x0000000100517b4c main + 164
6: 6   dyld                                0x000000019ef7ff28 start + 2236
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
❯ echo $MOOSE_ADFPARSER_JIT_INCLUDE
/Users/milljm/mambaforge3/envs/prebuilt/moose/include/moose/ADRealMonolithic.h
❯ file $MOOSE_ADFPARSER_JIT_INCLUDE
/Users/milljm/mambaforge3/envs/prebuilt/moose/include/moose/ADRealMonolithic.h: c program text, ASCII text
```
[log.txt](https://github.com/idaholab/moose/files/12457682/log.txt)

Reply:
This error is super weird IMO. The two declarations are in the _same_ file. I wonder if there is a missing preprocessor symbol that switches between them.

Reply:
> I'm just observing a structural problem that needs to be addressed IMO. The documentation was quite apparently either not enough, or not discoverable enough. Design docs are supposed to prevent these types of things from happening.
>
> I suppose we could link explicitly to the relevant design doc from an error message (ADFParser::JITCompile() failed. Evaluation not possible. Check JITCompilation.md to see if the environment is set up correctly.).

Just commenting quickly on this. Yea there are some flaws. Adding the relevant markdown file in the error would definitely make it more likely that we end up looking at it. I did test searching for JIT compilation on the website and the markdown file came up at the top with a couple different searches, so we probably should have come across it this time even without the explicit reference in the error. But sometimes unless I'm explicitly told something I may be too lazy to do the work 😦